BlueBrain/Search
# [BBS-293] Migrate spaCy 2.x -> 3.x

Fixes #274.

## Description

### PR progress

The detailed scope is available here: https://github.com/BlueBrain/Search/issues/274#issuecomment-801108330.
See especially the section "Out of scope" for links to follow-up work.

### PR changes

- Removed the installation of Prodigy in `pipelines/ner/Dockerfile`. Now `spacy train` is used.
- Removed the patch from #268. Patched issue fixed in spaCy >= 3.0.4.
- Migrated `add_pipe(EntityRuler, ...)` to spaCy 3. Now `add_pipe(...)` has a [new API](https://spacy.io/api/language#add_pipe).
- Migrated the NER training to `spacy train` and spaCy 3. Now Prodigy is no more needed to train NER models.
- Convert `.jsonl` annotations to `.spacy` files for spaCy 3. See #308 for migrating to `.spacy` permanently.
- Use the default configuration of spaCy 3 for NER training. See #309 for using the `config.cfg` from Prodigy.
- Upgraded `en_core_web_sm` and `scispaCy` models for spaCy 3.
- Upgraded `spaCy` to >= 3.0.4 and scispaCy to 0.4.0.

### Note on Prodigy for spaCy 3

At the moment, `Prodigy` has not been release for spaCy 3 yet.
The progress on the release of the new Prodigy could be followed [here](https://support.prodi.gy/t/timeline-for-spacy-3-integration/3529/) and [here](https://support.prodi.gy/t/prodigy-nightly-spacy-v3-support-ui-for-overlapping-spans-improved-feeds-more/3861).

### Note on the NER performances

The base models from `scispaCy` need to be upgraded (v0.2.5 to v0.4.0) to be loadable with `spaCy 3`.

Improving the performances of the trained models is **not** part of the PR. See #309, #294, and #295 instead.

There are performance changes compared to `master`(scores for 86844fe0fa3d9731c5bb0b4f9b43a3eeb57fe80d):

- According to the F1 score from `pipeline/ner/eval.py`:
  - 8 entities have an increase, especially `cell_compartment` (+12 %), `organism` (+7.2 %), and `drug` (+ 5.8 %)
  - 1 entity has a negligible decrease: `protein` (-0.73 %)
  - 1 entity has a concerning decrease: `pathway` (-19 %)

- According to the F1 score from `Prodigy` or `spaCy`:
  - 5 entities have an increase, especially `organism` (+31 %)
  - 1 entity has a negligible decrease: `pathway` (-0.58 %)
  - 3 entities have a concerning decrease: `cell_compartment` (-8.7 %), `protein` (-7.0 %), and `drug` (-2.5 %)

There are probably **issues** with 3 of the 5 entities mentioned above. Indeed:
- `pathway` has a catastrophic decrease according to eval.py (-19 %) but not spaCy (- 0.58 %)
- `cell_compartment` has a huge increase according to eval.py (+12 %) but has the opposite according to spaCy (-8.7 %)
- `drug` has an increase according to eval.py (+5.8 %) but has a decrease according to spaCy (-2.5%)

The changes seem to be caused by data sampling and domain adaptation issues. See #321, 3rd and 4th points.

Here are the full changes according to the F1 score from `pipeline/ner/eval.py`:

entity                     | before | now | delta | % |
------------------ | ------: | ---: | ----: | --: | 
cell_compartment | 0.65 | 0.72 | 0.08 | 12 | 
cell_type                | 0.64 | 0.64 | 0.00 | 0.0 | 
chemical                | 0.54 | 0.55 | 0.00 | 0.4 | 
disease                  | 0.69 | 0.70 | 0.01 | 1.3 | 
drug                       | 0.60 | 0.64 | 0.03 | 5.8 | 
organ                     | 0.53 | 0.55 | 0.02 | 3.7 | 
organism               | 0.54 | 0.58 | 0.04 | 7.2 | 
pathway                 | 0.58 | 0.47 | -0.11 | -19 | 
protein                   | 0.53 | 0.53 | -0.00 | -0.73 | 

Here are the full changes according to the F1 score from `Prodigy` or `spaCy`:

entity                     | before | now | delta | % |
------------------ | ------: | ---: | ----: | --: | 
cell_compartment | 0.88 | 0.81 | -0.08 | -8.7 | 
cell_type                | 0.77 | 0.78 | 0.01 | 1.7 | 
chemical                | 0.57 | 0.63 | 0.06 | 10.5 | 
disease                  | 0.88 | 0.92 | 0.04 | 4.6 | 
drug                       | 0.77 | 0.75 | -0.02 | -2.5 | 
organ                     | 0.81 | 0.85 | 0.04 | 5.0 | 
organism               | 0.68 | 0.89 | 0.21 | 31 | 
pathway                 | 0.84 | 0.84 | -0.00 | -0.58 | 
protein                   | 0.80 | 0.74 | -0.06 | -7.0 | 

## Note on reproducibility

The order of lines in `model-best/vocab/strings.json` changes between `dvc repro -f` calls.
This leads to changes in `pipeline/ner/dvc.lock` where the NER models are declared.

Otherwise, the outputs of the other parts of the pipeline (`convert_annotations_*`, `add_er_*`, `eval_*`) do not change .
This therefore implies that the performances of the trained models do not change.

A patch has been applied in d8b9f7407641f977a0091d61788a96c97b48fa4a for ordering `strings.json` deterministically.
See also #327 for removing the patch when spaCy will have fixed it upstream.

## How to test?

1. The following should execute without errors:

```
git clone https://github.com/BlueBrain/Search
git checkout bbs_293

docker build <options> -f data_and_models/pipelines/ner/Dockerfile -t <image> .
docker run -it -rm <options> --name <container> <image>

dvc pull data_and_models/annotations/ner/*.dvc

cd data_and_models/pipelines/ner/
# This takes around 30 mins.
dvc repro -f
```

2. The following, executed after the block above, should return nothing:

```
git diff
```

## Checklist

- [ ] All checkable items from https://github.com/BlueBrain/Search/issues/274#issuecomment-801108330 are checked.


## Tests failing

1. `add_pipe` API change:
```
ValueError: [E966] `nlp.add_pipe` now takes the string name of the registered component factory, not a callable component. Expected string, but got <spacy.pipeline.ner.EntityRecognizer object at 0x151ece910>
 - If you created your component with `nlp.create_pipe('name')`: remove nlp.create_pipe and call `nlp.add_pipe('name')` instead.
 - If you passed in a component like `TextCategorizer()`: call `nlp.add_pipe` with the string name instead, e.g. `nlp.add_pipe('textcat')`.
 - If you're using a custom component: Add the decorator `@Language.component` (for function components) or `@Language.factory` (for class components / factories) to your custom component and assign it a name, e.g. `@Language.component('your_name')`. You can then run `nlp.add_pipe('your_name')` to add it to the pipeline.
```

Change of syntax with an `entity_ruler`:
`spaCy<3`:
```
er = spacy.pipeline.EntityRuler(model, patterns=self.to_list())
model.add_pipe(er, **add_pipe_kwargs)
```
`spaCy>=3`:
```
er = model.add_pipe("entity_ruler",
                    config={'validate': True, **add_pipe_kwargs})
er.add_patterns(self.to_list())
```

2. `deepcopy` of spacy model: 
```
TypeError: self.c cannot be converted to a Python object for pickling
```
from `tests/test_mining/test_attribute.py` line 1225: 
``` 
extractor.ee_model = deepcopy(extractor.ee_model)
```
